Fix false skipping of folder-seperator

dan-42 opened the pull request:
@danmar sorry I planned top update this PR more and also do better research where and when this is the issue.
So in my test-example I generate a ` compile_commands.json` via CMake  ` set(CMAKE_EXPORT_COMPILE_COMMANDS ON) ` that looks like this:
```json
[
{
  "directory": "C:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug",
  "command": "C:\\PROGRA~2\\MICROS~1\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1427~1.291\\bin\\HostX64\\x64\\cl.exe  /nologo /TP -IC:\\Users\\dan\\git\\test-cppcheck\\mylib\\src /DWIN32 /D_WINDOWS /GR /EHsc /Zi /Ob0 /Od /RTC1 -MDd -std:c++17 /Fomylib\\CMakeFiles\\mylib.dir\\src\\foobar\\mylib.cpp.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\dan\\git\\test-cppcheck\\mylib\\src\\foobar\\mylib.cpp",
  "file": "C:\\Users\\dan\\git\\test-cppcheck\\mylib\\src\\foobar\\mylib.cpp"
},
{
  "directory": "C:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug",
  "command": "C:\\PROGRA~2\\MICROS~1\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1427~1.291\\bin\\HostX64\\x64\\cl.exe  /nologo /TP -IC:\\Users\\dan\\git\\test-cppcheck\\myapp\\src -Imyapp -IC:\\Users\\dan\\git\\test-cppcheck\\mylib\\src /DWIN32 /D_WINDOWS /GR /EHsc /Zi /Ob0 /Od /RTC1 -MDd -std:c++17 /Fomyapp\\CMakeFiles\\myapp.dir\\src\\main.cpp.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\dan\\git\\test-cppcheck\\myapp\\src\\main.cpp",
  "file": "C:\\Users\\dan\\git\\test-cppcheck\\myapp\\src\\main.cpp"
}
]
```


and the project has the following structure
```sh
/project
├── CMakeLists.txt
├── mylib/
│     ├── CMakeLists.txt
│     └── src
│              └── foobar <-- namespace
│                     ├── mylib.hpp  <-- has macro MYLOG(msg)
│                     └── mylib.cpp
└── myapp/
        ├── CMakeLists.txt
        └── src/
              └── main.cpp
```

Calling cppcheck with ` cppcheck --project=./compile_commands.json --std=c++17 --language=c++ --enable=all --verbose `
```sh
$ cppcheck --project=./compile_commands.json --std=c++17 --language=c++ --enable=all --verbose
Checking C:\Users\dan\git\test-cppcheck\mylib\src\foobar\mylib.cpp ...
Defines:WIN32=1;_WINDOWS=1
Undefines:
Includes: -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/C:Usersdangittest-cppcheckmylibsrc/
Platform:win64
Checking C:\Users\dan\git\test-cppcheck\mylib\src\foobar\mylib.cpp: WIN32=1;_WINDOWS=1...
1/2 files checked 50% done
Checking C:\Users\dan\git\test-cppcheck\myapp\src\main.cpp ...
Defines:WIN32=1;_WINDOWS=1
Undefines:
Includes: -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/C:Usersdangittest-cppcheckmyappsrc/ -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/C:Usersdangittest-cppcheckmylibsrc/ -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/myapp/
Platform:win64
Checking C:\Users\dan\git\test-cppcheck\myapp\src\main.cpp: WIN32=1;_WINDOWS=1...
2/2 files checked 100% done
C:\Users\dan\git\test-cppcheck\mylib\src\foobar\mylib.cpp:20:0: style: The function 'log' is never used. [unusedFunction]

^
nofile:0:0: information: Cppcheck cannot find all the include files. Cppcheck can check the code without the include files found. But the results will probably be more accurate if all the include files are found. Please check your project's include directories and add all of them as include directories for Cppcheck. To see what files Cppcheck cannot find use --check-config. [missingIncludeSystem]

```

note the line 
` Includes: -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/C:Usersdangittest-cppcheckmylibsrc/ `

this should be 
` Includes: -I/C:Users/dan/git/test-cppcheck/mylib/src/ `


I think this is only windows related, but still I think I'm not the first who want to use `cppcheck` with `CMake` on windows.

I think it does not  matter but I'm using the following tooling:
 * Windows10
 * msvc 2019 x64
 * Ninja 1.10.0
 * CMake 3.18.1
 * cppcheck 2.1
 * IDE: QtCreator 4.12.4


in addition I found this open topic on Ninja 
https://github.com/ninja-build/ninja/issues/869

danmar commented:
hello. hmm that is an interesting find.

danmar commented:
I am guessing than an include with spaces will look like?

    -I\\\"C:\some path\\\"

so do we need some special string handling..

dan-42 commented:
@danmar sorry I planned top update this PR more and also do better research where and when this is the issue.
So in my test-example I generate a ` compile_commands.json` via CMake  ` set(CMAKE_EXPORT_COMPILE_COMMANDS ON) ` that looks like this:
```json
[
{
  "directory": "C:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug",
  "command": "C:\\PROGRA~2\\MICROS~1\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1427~1.291\\bin\\HostX64\\x64\\cl.exe  /nologo /TP -IC:\\Users\\dan\\git\\test-cppcheck\\mylib\\src /DWIN32 /D_WINDOWS /GR /EHsc /Zi /Ob0 /Od /RTC1 -MDd -std:c++17 /Fomylib\\CMakeFiles\\mylib.dir\\src\\foobar\\mylib.cpp.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\dan\\git\\test-cppcheck\\mylib\\src\\foobar\\mylib.cpp",
  "file": "C:\\Users\\dan\\git\\test-cppcheck\\mylib\\src\\foobar\\mylib.cpp"
},
{
  "directory": "C:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug",
  "command": "C:\\PROGRA~2\\MICROS~1\\2019\\COMMUN~1\\VC\\Tools\\MSVC\\1427~1.291\\bin\\HostX64\\x64\\cl.exe  /nologo /TP -IC:\\Users\\dan\\git\\test-cppcheck\\myapp\\src -Imyapp -IC:\\Users\\dan\\git\\test-cppcheck\\mylib\\src /DWIN32 /D_WINDOWS /GR /EHsc /Zi /Ob0 /Od /RTC1 -MDd -std:c++17 /Fomyapp\\CMakeFiles\\myapp.dir\\src\\main.cpp.obj /FdTARGET_COMPILE_PDB /FS -c C:\\Users\\dan\\git\\test-cppcheck\\myapp\\src\\main.cpp",
  "file": "C:\\Users\\dan\\git\\test-cppcheck\\myapp\\src\\main.cpp"
}
]
```


and the project has the following structure
```sh
/project
├── CMakeLists.txt
├── mylib/
│     ├── CMakeLists.txt
│     └── src
│              └── foobar <-- namespace
│                     ├── mylib.hpp  <-- has macro MYLOG(msg)
│                     └── mylib.cpp
└── myapp/
        ├── CMakeLists.txt
        └── src/
              └── main.cpp
```

Calling cppcheck with ` cppcheck --project=./compile_commands.json --std=c++17 --language=c++ --enable=all --verbose `
```sh
$ cppcheck --project=./compile_commands.json --std=c++17 --language=c++ --enable=all --verbose
Checking C:\Users\dan\git\test-cppcheck\mylib\src\foobar\mylib.cpp ...
Defines:WIN32=1;_WINDOWS=1
Undefines:
Includes: -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/C:Usersdangittest-cppcheckmylibsrc/
Platform:win64
Checking C:\Users\dan\git\test-cppcheck\mylib\src\foobar\mylib.cpp: WIN32=1;_WINDOWS=1...
1/2 files checked 50% done
Checking C:\Users\dan\git\test-cppcheck\myapp\src\main.cpp ...
Defines:WIN32=1;_WINDOWS=1
Undefines:
Includes: -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/C:Usersdangittest-cppcheckmyappsrc/ -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/C:Usersdangittest-cppcheckmylibsrc/ -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/myapp/
Platform:win64
Checking C:\Users\dan\git\test-cppcheck\myapp\src\main.cpp: WIN32=1;_WINDOWS=1...
2/2 files checked 100% done
C:\Users\dan\git\test-cppcheck\mylib\src\foobar\mylib.cpp:20:0: style: The function 'log' is never used. [unusedFunction]

^
nofile:0:0: information: Cppcheck cannot find all the include files. Cppcheck can check the code without the include files found. But the results will probably be more accurate if all the include files are found. Please check your project's include directories and add all of them as include directories for Cppcheck. To see what files Cppcheck cannot find use --check-config. [missingIncludeSystem]

```

note the line 
` Includes: -IC:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug/C:Usersdangittest-cppcheckmylibsrc/ `

this should be 
` Includes: -I/C:Users/dan/git/test-cppcheck/mylib/src/ `


I think this is only windows related, but still I think I'm not the first who want to use `cppcheck` with `CMake` on windows.

I think it does not  matter but I'm using the following tooling:
 * Windows10
 * msvc 2019 x64
 * Ninja 1.10.0
 * CMake 3.18.1
 * cppcheck 2.1
 * IDE: QtCreator 4.12.4

dan-42 commented:
in addition I found this open topic on Ninja 
https://github.com/ninja-build/ninja/issues/869

dan-42 commented:
@danmar 
I updated the PR and now have also a fix for the issue.
I did some refactoring and added a unit-test to test my issue.

All tests pass, except travis which failed due to an internal error

I would love to have some feedback on the changes.

danmar commented on lib/importproject.cpp:
```diff
@@ -232,111 +232,165 @@ static std::string readUntil(const std::string &command, std::string::size_type
+static std::string parseTillNextCommandOpt(const std::string& command, std::string::size_type *pos)
+{
+    std::string ret{};
+    while (*pos < command.size() && command[*pos] != '/' && command[*pos] != '-' )
```
hmm.. not your code. But as far as I see this loop can be written as:

    *pos = command.find_first_not_of("/-", *pos);

dan-42 replied:
good catch, I will change it

danmar commented on lib/importproject.cpp:
```diff
@@ -232,111 +232,165 @@ static std::string readUntil(const std::string &command, std::string::size_type
+
+static std::string parseTillNextCommandOpt(const std::string& command, std::string::size_type *pos)
+{
+    std::string ret{};
```
that `{}` is not needed.

dan-42 replied:
Just a habit, I'll clean it up

danmar commented on lib/importproject.cpp:
```diff
@@ -217,13 +217,13 @@ ImportProject::Type ImportProject::import(const std::string &filename, Settings
     while (*pos < command.size() && (str || !std::strchr(until, command[*pos]))) {
-        if (command[*pos] == '\\')
-            ++*pos;
+        if (skipBackSlash && command[*pos] == '\\')
```
hmm.. somehow we want to skip escape backslashes but we do not want to skip path separators. I wonder what would happen if you check if there comes a `"` after then backslashes and if so removes them. It seems strange to parse different compiler options differently.

dan-42 replied:
But because the are different we need to parse them differently. Checking for ` " ` will not work because the includes are not necessarily wrapped by ` " `  you see it in the added unit-test

danmar replied:
sounds strange to me .. I'll need to understand how they are different.. can you show a -D and -I flag with similar information to show how they are different? For instance it would be interesting to see a comparison of `-I"c:\some path"` and `-DFOO="c:\some path"`

dan-42 replied:
@danmar that is a good example you already provided.
we first parse for the define e.g. ` -D ` then parse until we hit `=` or space and the we parse again the value of the define, if there is one.
And the first part is different to an include you do not have a `-DC:\\some path`, the path is only in the value.
So parsing the value of a define and parsing an include can be the same

danmar commented on lib/importproject.cpp:
```diff
@@ -232,111 +232,165 @@ static std::string readUntil(const std::string &command, std::string::size_type
+    }
+    (*pos)++;
+    if (*pos >= command.size())
+    {
```
stylistic nit; we do not usually use redundant `{}` around a single statement. And we usually do not write `{` on its own line. Well this would be corrected by our `runastyle` tool later but it looks nice if it follows normal cppcheck syntax directly.

danmar replied:
> And we usually do not write { on its own line. Well this would be corrected by our runastyle tool later but it looks nice if it follows normal cppcheck syntax directly.

You can ignore that for now. It will be fixed by runastyle later.

But well feel free to try out the runastyle script now immediately. You need to install astyle 3.0.1 , the exact version is very important to avoid formatting conflicts.

dan-42 replied:
Thx, yes I'm aware that the style is not 100% correct, on some places I also only used 2 instead of 4 spaces and so on.
I'm just used to different styling...
But I did not want to waste to much time in cleanup when the general solution would not fit.

So I will clean it up, promise🤞

danmar commented on lib/importproject.cpp:
```diff
@@ -232,111 +232,165 @@ static std::string readUntil(const std::string &command, std::string::size_type
+
+void ImportProject::FileSettings::parseCommandDefine(const std::string& command, std::string::size_type *pos, std::string& defs)
+{
+    const std::string fval = readUntil(command, pos, " =");
```
isn't it enough to write:

    defs += readUntil(command, pos, " ") + ";";

dan-42 replied:
I'll clean it up

danmar commented on lib/importproject.cpp:
```diff
@@ -232,111 +232,165 @@ static std::string readUntil(const std::string &command, std::string::size_type
-            }
+        if(opt=="D")
+        {
+          parseCommandDefine(command, &pos, defs);
```
I like that you broke out some functions here.

danmar commented on lib/importproject.cpp:
```diff
@@ -232,111 +232,165 @@ static std::string readUntil(const std::string &command, std::string::size_type
+        }
+        else if(opt=="municode")
+        {
+          defs += "UNICODE";
```
I'd prefer a single line:

    defs += "UNICODE;";

dan-42 replied:
👍

danmar commented on test/testimportproject.cpp:
```diff
@@ -141,6 +141,25 @@ class TestImportProject : public TestFixture {
+        const char json[] =
+            "[{"
+            "\"directory\": \"C:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug\","
+            "\"command\": \"C:\\\\PROGRA~2\\\\MICROS~1\\\\2019\\\\COMMUN~1\\\\VC\\\\Tools\\\\MSVC\\\\1427~1.291\\\\bin\\\\HostX64\\\\x64\\\\cl.exe  /nologo /TP -IC:\\\\Users\\\\dan\\\\git\\\\test-cppcheck\\\\mylib\\\\src /DWIN32 /D_WINDOWS /GR /EHsc /Zi /Ob0 /Od /RTC1 -MDd -std:c++17 /Fomylib\\\\CMakeFiles\\\\mylib.dir\\\\src\\\\foobar\\\\mylib.cpp.obj /FdTARGET_COMPILE_PDB /FS -c C:\\\\Users\\\\ ...
```
It seems to me it would be interesting with a include path with a space also. do you know if we have that somewhere? maybe that would just generate some "~" path then it's not very interesting.

dan-42 replied:
I agree I will add another test for that, folder-names with spaces are quite common on windows

dan-42 replied:
And I found an issue with folders and spaces :-)
added a fix for that

danmar commented on test/testimportproject.cpp:
```diff
@@ -141,6 +141,25 @@ class TestImportProject : public TestFixture {
+    void importCompileCommands5() const {
+        const char json[] =
+            "[{"
+            "\"directory\": \"C:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug\","
```
hmm.. we should consider writing these json testcases with raw strings instead. The escaped escapes makes it hard to read. well.. you followed the convention so I will merge this if you keep the code as it is. We could look at that afterwards.

dan-42 replied:
I'm  bit confused, how would you write it as a "raw string"?
The generated file file really has `\\` and to have it  in code both need to be escaped which makes it `\\\\`

But I'm happy to change it, I just do not understand what you mean :-)

danmar replied:
you should be able to copy paste the raw json, for instance:

    R"([{
              "directory": "C:/Users/dan/git/build-test-cppcheck-Desktop_Qt_5_15_0_MSVC2019_64bit-Debug",
              "command": "C:\\PROGRA~2\\MICROS~1\\201....."
              .....
         }])";

danmar replied:
it would have felt better with raw strings

danmar reviewed (commented):
looks good! minor nits.

dan-42 commented:
@danmar thx for your feedback, I'll try to address them soon

dan-42 commented:
@danmar I cleaned it up and addressed all your feedback.

Thank you for your feedback, now cppcheck can run almost over the whole project I have 🥳

amai2012 commented on lib/importproject.cpp:
```diff
@@ -327,13 +310,14 @@ void ImportProject::FileSettings::parseCommandInclude(const std::string& command
-  const bool skipBackSlash = false;
-  const std::string isystem = readUntil(command, pos, " ", skipBackSlash);
-  systemIncludePaths.push_back(isystem);
+    const bool skipBackSlash = false;
```
Where does this reformatting come from?

dan-42 replied:
@amai2012 I used ` astyle` and  the `runastyle` script

dan-42 commented:
@danmar Are there any pending tasks that are needed from me to have this merged?

danmar commented:
I would like to understand why `-I` and `-D` should have different parsing with `skipBackslashes`. Would it be possible to show the output for some `-I"c:\some path"` and `-DFOO="c:\some path"`. Spontaneously it seems weird to me that cmake encodes these flags in different ways.

danmar commented:
From a test I made:

    -DFILESDIR=\"\\\"/home/danielm/cppcheck 2\\\"\" -I\"/home/danielm/cppcheck 2/build/externals/tinyxml\"

I can use `\` in folder names in linux, for instance I tried to rename "cppcheck" to "c:\some path" ... but cmake did not like that it thought the `\` was a path separator

dan-42 commented:
@danmar I wanted to answer you quickliy but then I did some tests and it will take more time to properly check it.
I might still have some issues in my implementation.

In general I agree the value of the definition should be parsed the same way as an include.
So maybe I'll find the time again this evening.

dan-42 commented:
@danmar I cleaned it up and fixed some issues I found.

Now `` define values `` and `` includes `` are now parsed the same.
But I figured that the existing unit-test for defines with escaping (`` -DFILESDIR=\\\\\\\"/usr/local/share/Cppcheck\\\\\\\" ``)
used  wrong escaping.
But since the old code skipped all backslashes this was not discovered.

Parsing `` define values `` or `` includes `` do not strip any quotation marks, this is done while processing the includes separately.
This can easily be added also for defines with `` Path::removeQuotationMarks() `` But this might change existing behavior. on the `` define values `` .

dan-42 commented:
@danmar The CI worked, 
only ` CI-unixish` failed due to some upload issues, so nothing related to my code.

danmar commented:
> But I figured that the existing unit-test for defines with escaping (-DFILESDIR=\\\\\\\"/usr/local/share/Cppcheck\\\\\\\")
used wrong escaping.

Yes.. we need to start using raw strings. :-)

danmar commented:
Well.. I merge this ... then convert to raw strings.. and then let's hope this will work..

danmar commented:
I have converted to raw strings..

Can you please look at `TestImportProject::importCompileCommands7`. The parsing of `-DFILESDIR=\"\\\"/home/danielm/cppcheck 2\\\"\"` does not work perfectly.

dan-42 commented:
@danmar thx for merging and yes I will look into it

dan-42 commented:
OK Well here it was quicker then I thought it would be...

### So the test string
`` -DFILESDIR=\"\\\"/home/danielm/cppcheck 2\\\"\" -DTEST1 -DTEST2=2  ``
is parsed to:
``  FILESDIR=\"\\"/home/danielm/cppcheck;TEST1=1;TEST2=2 ``

which kinda makes sens...
Looking at it  in the raw-string
I wonder how do you expect to match to 
`` \"/home/danielm/cppcheck 2\" ``

why should the first `` \"\\  `` be stripped away?  
The ending is only stripped correctly because the overall parsing does skip/ignore it. But `` readUntil() `` does not do it.

### My solution would be to fixing the test

"command": "gcc -DFILESDIR=\"/home/danielm/cppcheck 2\" -I\"/home/danielm/cppcheck 2/build/externals/tinyxml\" -DTEST1 -DTEST2=2 -o /tmp/src.o -c /tmp/src.c",

But maybe this is a real use case which needs to be handled?

danmar commented:
> why should the first \"\\ be stripped away?

ok that was probably wrong.. you can probably keep that. Just ensure that it does  not stop at the space.

> But maybe this is a real use case which needs to be handled?

Yes I assume you can reproduce it. Go to the cppcheck repo. Try this command: `cmake -DFILESDIR="/home/danielm/cppcheck 2" ..`

danmar commented:
For information.. with this command: `cmake -DFILESDIR="c:\some path" ..`

For me, the compile commands file says:

{
  "directory": "/home/danielm/cppcheck/b/externals/tinyxml",
  "command": "/usr/bin/c++  -DFILESDIR=\"\\\"c:\\\\some path\\\"\" -D_GLIBCXX_DEBUG -I/home/danielm/cppcheck/b/externals/tinyxml -I/home/danielm/cppcheck/externals/tinyxml  -g   -Wcast-qual -Wno-deprecated-declarations -Wfloat-equal -Wmissing-declarations -Wmissing-format-attribute -Wno-long-long -Woverloaded-virtual -Wpacked -Wredundant-decls -Wundef -Wno-shadow -Wno-missing-field-initializers -Wno-missing-braces -Wno-sign-compare -Wno-multichar -Wno-maybe-uninitialized -Wno-suggest-attribute=format -std=gnu++11 -o CMakeFiles/tinyxml_objs.dir/tinyxml2.cpp.o -c /home/danielm/cppcheck/externals/tinyxml/tinyxml2.cpp",
  "file": "/home/danielm/cppcheck/externals/tinyxml/tinyxml2.cpp"
},

I assume that everything inside the string is double escaped..
